File: web/ios/Omnigent/DatabricksOAuthAttempt.swift
import CryptoKit
import Foundation
import Security

/// Transient material for one authorization-code flow; never persisted or sent to JavaScript.
struct DatabricksOAuthAttempt: Sendable {
  let configuration: DatabricksOAuthConfiguration
  let credentialScope: DatabricksCredentialScope
  var workspaceOrigin: URL { credentialScope.workspaceOrigin }
  let state: String
  let verifier: String

  init(workspaceURL: URL, configuration: DatabricksOAuthConfiguration) throws {
    credentialScope = try DatabricksCredentialScope(
      workspaceURL: workspaceURL, configuration: configuration)
    self.configuration = configuration
    state = try Self.randomValue()
    verifier = try Self.randomValue()
  }

  var authorizationURL: URL {
    var components = URLComponents(
      url: workspaceOrigin.appendingPathComponent("oidc/v1/authorize"),
      resolvingAgainstBaseURL: false)!
    components.queryItems = [
      URLQueryItem(name: "client_id", value: configuration.clientID),
      URLQueryItem(name: "redirect_uri", value: configuration.redirectURL.absoluteString),
      URLQueryItem(name: "response_type", value: "code"),
      URLQueryItem(name: "scope", value: Self.scope),
      URLQueryItem(name: "state", value: state),
      URLQueryItem(name: "code_challenge", value: Self.challenge(for: verifier)),
      URLQueryItem(name: "code_challenge_method", value: "S256"),
    ]
    // OAuth servers decode query parameters as form data, where a literal + is a space.
    components.percentEncodedQuery = components.percentEncodedQuery?.replacingOccurrences(
      of: "+", with: "%2B")
    return components.url!
  }

  func authorizationCode(from callback: URL) throws -> String {
    guard configuration.matchesCallback(callback),
      let items = URLComponents(url: callback, resolvingAgainstBaseURL: false)?.queryItems
    else { throw DatabricksOAuthError.invalidCallback }
    let states = items.filter { $0.name == "state" }
    let codes = items.filter { $0.name == "code" }
    let errors = items.filter { $0.name == "error" }
    guard states.count == 1, states.first?.value == state else {
      throw DatabricksOAuthError.invalidCallback
    }
    if !errors.isEmpty {
      guard errors.count == 1, codes.isEmpty, let error = errors.first?.value, !error.isEmpty
      else { throw DatabricksOAuthError.invalidCallback }
      throw error == "access_denied"
        ? DatabricksOAuthError.authorizationDenied : DatabricksOAuthError.authorizationFailed
    }
    guard codes.count == 1, let code = codes.first?.value, !code.isEmpty else {
      throw DatabricksOAuthError.invalidCallback
    }
    return code
  }

  func tokenRequest(code: String) -> URLRequest {
    DatabricksOAuthClient.tokenRequest(
      for: credentialScope,
      fields: [
        ("grant_type", "authorization_code"),
        ("redirect_uri", configuration.redirectURL.absoluteString), ("scope", Self.scope),
        ("code_verifier", verifier), ("code", code),
      ])
  }

  static func challenge(for verifier: String) -> String {
    base64URL(Data(SHA256.hash(data: Data(verifier.utf8))))
  }

  private static let scope = "all-apis offline_access"

  private static func randomValue() throws -> String {
    var bytes = [UInt8](repeating: 0, count: 32)
    guard SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes) == errSecSuccess else {
      throw DatabricksOAuthError.randomGenerationFailed
    }
    return base64URL(Data(bytes))
  }

  private static func base64URL(_ data: Data) -> String {
    data.base64EncodedString().replacingOccurrences(of: "+", with: "-")
      .replacingOccurrences(of: "/", with: "_").replacingOccurrences(of: "=", with: "")
  }

}

File: web/android/app/src/main/java/ai/omnigent/android/DatabricksOAuthAttempt.kt
package ai.omnigent.android

import org.json.JSONObject
import java.net.URI
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64

/** Transient material for one authorization-code flow. */
data class DatabricksOAuthAttempt(
    val configuration: DatabricksOAuthConfiguration,
    val credentialScope: DatabricksCredentialScope,
    val state: String,
    val verifier: String,
) {
    val authorizationUri: URI
        get() {
            val fields =
                mutableListOf(
                    "client_id" to configuration.clientId,
                    "redirect_uri" to configuration.redirectUri.toString(),
                    "response_type" to "code",
                    "scope" to SCOPE,
                    "state" to state,
                    "code_challenge" to challenge(verifier),
                    "code_challenge_method" to "S256",
                )
            credentialScope.workspaceId?.let { fields += "o" to it }
            val query = fields.joinToString("&") { (key, value) -> "$key=${formEncode(value)}" }
            return URI("${credentialScope.workspaceOrigin}/oidc/v1/authorize?$query")
        }

    fun authorizationResponse(callback: URI): AuthorizationResponse {
        if (!configuration.matchesCallback(callback)) {
            throw DatabricksOAuthException.InvalidCallback()
        }
        val items =
            try {
                queryItems(callback)
            } catch (_: Throwable) {
                throw DatabricksOAuthException.InvalidCallback()
            }
        val states = items.filter { it.first == "state" }
        val codes = items.filter { it.first == "code" }
        val errors = items.filter { it.first == "error" }
        if (states.size != 1 || states.single().second != state) {
            throw DatabricksOAuthException.InvalidCallback()
        }
        if (errors.isNotEmpty()) {
            if (errors.size != 1 || codes.isNotEmpty() || errors.single().second.isNullOrEmpty()) {
                throw DatabricksOAuthException.InvalidCallback()
            }
            if (errors.single().second == "access_denied") {
                throw DatabricksOAuthException.AuthorizationDenied()
            }
            throw DatabricksOAuthException.AuthorizationFailed()
        }
        val code = codes.singleOrNull()?.second
        if (code.isNullOrEmpty()) throw DatabricksOAuthException.InvalidCallback()
        val issuers = items.filter { it.first == "iss" }
        val issuer =
            if (issuers.isEmpty()) {
                DatabricksOAuthIssuer(URI("${credentialScope.workspaceOrigin}/oidc"))
            } else {
                val value = issuers.singleOrNull()?.second
                if (value.isNullOrEmpty()) throw DatabricksOAuthException.InvalidIssuer()
                try {
                    DatabricksOAuthIssuer(URI(value))
                } catch (error: DatabricksOAuthException.InvalidIssuer) {
                    throw error
                } catch (_: Throwable) {
                    throw DatabricksOAuthException.InvalidIssuer()
                }
            }
        return AuthorizationResponse(code, issuer)
    }

    data class AuthorizationResponse(
        val code: String,
        val issuer: DatabricksOAuthIssuer,
    )

    companion object {
        private const val SCOPE = "all-apis offline_access"
        private val secureRandom = SecureRandom()

        fun create(
            workspaceUrl: URI,
            configuration: DatabricksOAuthConfiguration,
            randomBytes: (Int) -> ByteArray = ::secureRandomBytes,
        ): DatabricksOAuthAttempt =
            DatabricksOAuthAttempt(
                configuration = configuration,
                credentialScope = DatabricksCredentialScope.from(workspaceUrl, configuration),
                state = mobileRedirectState(base64Url(randomBytes(32))),
                verifier = base64Url(randomBytes(32)),
            )

        fun challenge(verifier: String): String =
            base64Url(MessageDigest.getInstance("SHA-256").digest(verifier.toByteArray()))

        fun mobileRedirectState(nonce: String): String {
            val document =
                JSONObject()
                    .put("scheme", MOBILE_REDIRECT_SCHEME)
                    .put("nonce", nonce)
                    .toString()
            return Base64.getEncoder().encodeToString(document.toByteArray())
        }

        const val MOBILE_REDIRECT_SCHEME = "ai.omnigent.android"

        private fun secureRandomBytes(size: Int): ByteArray =
            ByteArray(size).also(secureRandom::nextBytes)

        private fun base64Url(bytes: ByteArray): String =
            Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)
    }
}
